File: packages/shared_types/src/schema.ts
/**
 * packages/shared-types/src/schema.ts
 * Single source of truth for entire system
 */

export interface SatelliteScene {
    id: string;                          // S2A_20240115T052941_N0509_R105_T44PKT
    constellation: 'SENTINEL_2' | 'LANDSAT_9' | 'PLANET_SCOPE';
    satelliteId: string;                 // S2A, S2B, LC09
    acquisitionTime: string;             // ISO 8601
    cloudCover: number;                  // 0-100
    bbox: [number, number, number, number]; // [minX, minY, maxX, maxY]
    resolution: number;                  // 10, 30, 3 meters
    sizeBytes: number;
    status: 'ARCHIVED' | 'PROCESSING' | 'READY' | 'FAILED';
    bands: ('B02' | 'B03' | 'B04' | 'B08' | 'B11' | 'B12')[];
    downloadUrl?: string;
    cogUrl?: string;                     // Cloud Optimized GeoTIFF
}

export interface Detection {
    id: string;                          // UUID v4
    sceneId: string;                     // Parent satellite scene
    zoneId: string;                      // vaigai_riverbed, etc.

    // Geometry
    geometry: {
        type: 'Polygon';
        coordinates: number[][][];
    };
    centerLat: number;
    centerLon: number;
    areaSqm: number;

    // AI Predictions
    confidence: number;                  // 0-1
    wasteType: 'FRESH_MSW' | 'CONSTRUCTION' | 'CHEMICAL' | 'LEACHATE' | 'LEGACY';
    growthRate?: number;                 // m²/day
    toxicityIndex?: number;              // 0-10

    // Temporal
    firstDetectedAt: string;
    lastVerifiedAt?: string;
    status: 'NEW' | 'VERIFIED' | 'FALSE_POSITIVE' | 'CLEANED' | 'EXPANDING';

    // Provenance
    modelVersion: string;
    processingPipeline: string;

    // Civic Integration
    municipalTicketId?: string;
    alertSent: boolean;
    groundPhotos?: string[];
}

export interface Zone {
    id: string;
    name: string;
    nameTamil: string;
    centerLat: number;
    centerLon: number;
    radiusM: number;
    riskLevel: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';
    municipalWard: string;
    policeStation: string;
    estimatedPopulation: number;
    lastInspectionDate?: string;
}

// Real-time streaming events
export interface StreamEvent {
    type: 'SCENE_ACQUIRED' | 'DETECTION_CREATED' | 'ALERT_TRIGGERED' | 'MODEL_UPDATED';
    timestamp: string;
    payload: unknown;
    priority: 1 | 2 | 3 | 4 | 5;
}

File: packages/shared_types/src/schema.py
# packages/shared-types/src/schema.py
# Python mirror of TypeScript schemas

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import Optional, List, Tuple, Any
from pydantic import BaseModel, Field

# Assuming geojson-pydantic is available or using a simple Polygon model
class Polygon(BaseModel):
    type: str = "Polygon"
    coordinates: List[List[List[float]]]

class Constellation(str, Enum):
    SENTINEL_2 = "SENTINEL_2"
    LANDSAT_9 = "LANDSAT_9"
    PLANET_SCOPE = "PLANET_SCOPE"

class WasteType(str, Enum):
    FRESH_MSW = "FRESH_MSW"
    CONSTRUCTION = "CONSTRUCTION"
    CHEMICAL = "CHEMICAL"
    LEACHATE = "LEACHATE"
    LEGACY = "LEGACY"

class Detection(BaseModel):
    id: str = Field(..., description="UUID v4")
    scene_id: str
    zone_id: str
    geometry: Polygon
    center_lat: Decimal = Field(..., max_digits=10, decimal_places=8)
    center_lon: Decimal = Field(..., max_digits=11, decimal_places=8)
    area_sqm: float
    confidence: float = Field(..., ge=0, le=1)
    waste_type: WasteType
    growth_rate: Optional[float] = None
    toxicity_index: Optional[float] = Field(None, ge=0, le=10)
    first_detected_at: datetime
    last_verified_at: Optional[datetime] = None
    status: str = "NEW"
    model_version: str
    processing_pipeline: str
    municipal_ticket_id: Optional[str] = None
    alert_sent: bool = False
    ground_photos: Optional[List[str]] = None
    
    class Config:
        from_attributes = True
